relayer/src/upgrade_chain.rs

```diff
@@ -0,0 +1,120 @@
+use bitcoin::hashes::core::time::Duration;
+use prost_types::Any;
+use thiserror::Error;
+use tracing::error;
+
+use ibc::ics02_client::client_state::AnyClientState;
+use ibc::ics02_client::height::Height;
+use ibc::ics24_host::identifier::{ChainId, ClientId};
+use ibc::{events::IbcEvent, ics07_tendermint::client_state::ClientState};
+use ibc_proto::cosmos::gov::v1beta1::MsgSubmitProposal;
+use ibc_proto::cosmos::upgrade::v1beta1::{Plan, SoftwareUpgradeProposal};
+
+use crate::chain::{Chain, CosmosSdkChain};
+use crate::config::ChainConfig;
+use crate::error::Error;
+
+#[derive(Debug, Error)]
+pub enum UpgradeChainError {
+    #[error("failed with underlying cause: {0}")]
+    Failed(String),
+
+    #[error("key error with underlying cause: {0}")]
+    KeyError(Error),
+
+    #[error(
+        "failed during a transaction submission step to chain id {0} with underlying error: {1}"
+    )]
+    SubmitError(ChainId, Error),
+}
+
+#[derive(Clone, Debug)]
+pub struct UpdatePlanOptions {
+    pub src_chain_config: ChainConfig,
+    pub dst_chain_config: ChainConfig,
+    pub src_client_id: ClientId,
+    pub amount: u64,
+    pub height_offset: u64,
+}
+
+pub fn build_and_send_upgrade_chain_message(
+    mut dst_chain: CosmosSdkChain, // the chain whose account is debited
+    src_chain: CosmosSdkChain,     // the chain where the transfer is sent
+    opts: &UpdatePlanOptions,
+) -> Result<Vec<IbcEvent>, UpgradeChainError> {
+    // build a proposal Plan
+    let upgrade_height = dst_chain
+        .query_latest_height()
+        .unwrap()
+        .add(opts.height_offset);
+
+    let client_state = src_chain
+        .query_client_state(&opts.src_client_id, Height::zero())
+        .unwrap();
+
+    let mut upgraded_client_state = ClientState::zero_custom_fields(client_state);
+    upgraded_client_state.latest_height = upgrade_height.increment();
+    upgraded_client_state.unbonding_period = Duration::from_secs(400 * 3600); // TODO add to options
+
+    let raw_client_state = AnyClientState::Tendermint(upgraded_client_state);
+    let plan = Plan {
+        name: "test".to_string(),
```

Comment:
Does this name matter or needs to be configured?

Reply:
No need to be configured. We can add later more options but I don't see a need right now.